qiskit/circuit/library/generalized_gates/permutation.py

```diff
@@ -75,9 +75,19 @@ def __init__(self,
         inner = QuantumCircuit(num_qubits, name=name)
 
         super().__init__(num_qubits, name=name)
-        for i in range(num_qubits):
-            if (pattern[i] != -1) and (pattern[i] != i):
-                inner.swap(i, int(pattern[i]))
-                pattern[pattern[i]] = -1
+        for i, j in _get_ordered_swap(pattern):
+            inner.swap(i, j)
+
         all_qubits = self.qubits
         self.append(inner, all_qubits)
+
+
+def _get_ordered_swap(permutation_in):
+    permutation = list(permutation_in[:])
+    swap_list = []
+    for i, val in enumerate(permutation):
+        if val != i:
+            j = permutation.index(i)
+            swap_list.append((i, j))
+            permutation[i], permutation[j] = permutation[j], permutation[i]
+    return swap_list[::-1]
```

Comment:
can you explain what this is doing (how the swaps are added)?
my thinking was that this problem could be approached as a sorting problem so doing something like Quicksort should give the minimal number of swaps to get the permutation.

Reply:
I'll put a description into the doc string. As far as efficiency I compared this method to another based on "cycle decomposition" which is supposed to be efficient and this was as efficient. Here is a comparison trying all permutations of a given width for `_get_ordered_swap`, cycle decomposition and [a quick sort implementation here](https://stackabuse.com/quicksort-in-python/). It seems sometimes quick sort does more swaps.

| `(width, method)`  |   `max(num_swaps)` |   `sum(swaps)` |
|:--------------------|-----------------------:|-----------------------:|
| (5, 'cycle_decomp') |                      4 |          326           |
| (5, 'ordered_swap') |                      4 |          326           |
| (5, 'quick_sort')   |                      4 |          328           |
| (6, 'cycle_decomp') |                      5 |         2556           |
| (6, 'ordered_swap') |                      5 |         2556           |
| (6, 'quick_sort')   |                      6 |         2620           |
| (7, 'cycle_decomp') |                      6 |        22212           |
| (7, 'ordered_swap') |                      6 |        22212           |
| (7, 'quick_sort')   |                      7 |        23096           |
| (8, 'cycle_decomp') |                      7 |       212976           |
| (8, 'ordered_swap') |                      7 |       212976           |
| (8, 'quick_sort')   |                      9 |       225504           |
| (9, 'cycle_decomp') |                      8 |            2.23934e+06 |
| (9, 'ordered_swap') |                      8 |            2.23934e+06 |
| (9, 'quick_sort')   |                     11 |            2.40672e+06 |

Reply:
nice, thanks for this detailed evaluation